@grafana/data: Matchers and Transforms

ryantxu opened the pull request:
This PR proposes two new key types aimed to let us manipulate DataFrame[] results in the frontend.  The APIs are designed to be easily driven by settings saved in json (ie, panel/dashboard configs).  The initial APIs are not focused on using these features from javascript -- though that should come in a later issue.

**Matchers** 
fieldMatcher and frameMatcher.  This is various config options that let you say if something should apply to a field or DataFrame:
```
export type FieldMatcher = (field: Field) => boolean;
export type FrameMatcher = (frame: DataFrame) => boolean;
```

**Transformers** 
This represents a function that will manipulate DataFrame[]:
```
export type DataTransformer = (data: DataFrame[]) => DataFrame[];
```
Transformations will support standard features like 'rows to columns', 'time series to rows' etc that already exist in the Table panel.


this PR does not aim to build the UI to configure the functions, but is designed with that in mind.

ryantxu committed:
add extension framework

ryantxu committed:
add filter transformer

torkelo commented:
I like where this is going!

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  GraphLegendEditor: use stats picker rather than switches (#16759)
  Feature: add cron setting for the ldap settings (#16673)
  Build: Disables gosec until identified performance problems (#16764)
  Chore: bump jQuery to 3.4.0 including prototype pollution vulnerability fix (#16761)
  elasticsearch: add 7.x version support (#16646)
  Provisioning: Add API endpoint to reload provisioning configs (#16579)
  Config: Show user-friendly error message instead of stack trace (#16564)
  Chore: Lowered implicit anys limit to 5954
  Feature: Enable React based options editors for Datasource plugins (#16748)
  sqlstore: use column name in order by (#16583)
  user friendly guide (#16743)
  Provisioning: Interpolate env vars in provisioning files (#16499)
  admin: add more stats about roles (#16667)
  Feature: Migrate Legend components to grafana/ui (#16468)
  playlist: fix loading dashboards by tag (#16727)
  CloudWatch: Use default alias if there is no alias for metrics (#16732)
  Provisioning: Support FolderUid in Dashboard Provisioning Config (#16559)

ryantxu committed:
more logging

ryantxu committed:
adding more tests

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  build: restore postgres integration tests (#16801)
  docs: explain correct access control model of GCS buckets (#16792)
  Chore: Fixed no implicit any Typescript errors (#16799)
  Feature: introduce LdapActiveSyncEnabled setting (#16787)
  Plugins: ReactPanelPlugin to VizPanelPlugin (#16779)
  UX: Improve Grafana usage for smaller screens  (#16783)
  ThresholdEditor: Minor style fix for smaller screens (#16791)
  Build: Use isolated modules for ts-jest (#16786)
  LDAP Refactoring to support syncronizing more than one user at a time. (#16705)
  build: removes unused vendored files
  (fix/explore): remove vertical-align looks better for long logs (#16736)
  Chore: bump jQuery to 3.4.0 in grafana/ui (#16781)
  Devenv: Updated home dashboard and added new influxdb test dashboard
  Chore: Lowered implicit anys limit to 5946
  RefreshPicker: minor design update (#16774)
  Streaming: support streaming and a javascript test datasource (#16729)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  Explore: Support user timezone (#16469)
  Plugins: rename vizPlugin to panelPlugin (#16802)
  Plugins: move app/feature/plugin properties into PluginMeta (#16809)
  Plugins: move PanelPluginMeta to grafana/ui (#16804)
  Plugins: move datasource specific meta out of the main meta type (#16803)
  updates changelog for 6.1.6
  Fix: Fetch histogram series from other api route (#16768)
  phantomjs: set web-security to true
  Chore: Lowered implicit anys limit to 5668

ryantxu committed:
make stats an extension

ryantxu committed:
make stats an extension

ryantxu committed:
test registry init

ryantxu committed:
first get a function, then call it

ryantxu committed:
merge master

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  FormLabel: allow any rather than just a string for tooltip (#16841)
  prometheus: fix regression after adding support for tracing headers (#16829)
  area/circleci: Speed up circleci build process for branches and pr (#16778)
  DataProxy: Restore Set-Cookie header after proxy request (#16838)
  docs: clarify page parameter version support for folder/dashboard search (#16836)
  Chore: revise some of the gosec rules (#16713)
  Refactor: consistant plugin/meta usage (#16834)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  Chore: Lowered implicit anys limit to 5623
  TestData: Add dashboards to testdata (#16855)
  Input Datasource: convert from angular config to react ConfigEditor (#16856)
  DataSources: minor typescript cleanups and comments (#16860)
  TestDataDatasource: Add config editor (#16861)
  App Plugins: support react pages and tabs (#16586)
  Add Windows MSI generation to build process (#16502)
  Datasources: add support for POST HTTP verb for InfluxDB (#16690)
  Add pattern validation in configs (#16837)
  Search: Enable filtering dashboards in search by current folder (#16790)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  CSV: escape quotes in toCSV  (#16874)
  Dashboard: Lazy load out of view panels (#15554)
  LDAP: Added reload endpoint for LDAP config (#15470)
  PluginsList: Removed icons and updated snapshots (#16872)
  Panels: Fixed issue with panel type change and data updates (#16871)
  Chore: fix modes for non-executable files (#16864)
  DataSourceSettings: Minor fix for uncontrolled input (#16863)

ryantxu committed:
merge master

ryantxu committed:
merge master

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master: (39 commits)
  GettingStarted: add key and remove ng-class (#17007)
  explore: add some extra time for angular query editors to update query (#16955)
  Explore: Align Explore with Dashboards and Panels (#16823)
  Explore: Fix empty result from datasource should render logs container (#16999)
  Explore: Fixes zoom exception in Loki/Graph (#16991)
  PanelEditor: Fix queries tab now showing, wrong skipDataQuery logic (#16994)
  DataSourceApi:  convert interface to abstract class (#16979)
  Panels: Fixed error panel tooltip (#16993)
  Docker: Prevent a permission denied error when writing files to the default provisioning directory (#16831)
  Notification: attempt to send notifications to all given email addresses (#16881)
  GettingStarted: convert to react panel plugin (#16985)
  Plugins: Remove dataFormats key and add skipDataQuery (#16984)
  AlertList: removed icon
  MetricsPanelCtrl: use shared queryRunner to support query execution (#16659)
  TableData: support name (#16983)
  Changelog: Typo guage -> gauge (#16982)
  TestData: stream via fetch (#16963)
  plugins: fix how datemath utils are exposed to plugins (#16976)
  NewDataSource: Updated page header title
  fix(prometheus): issue with click label to filter for recording rules in Explore
  ...

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  chore: remove x character in explore
  Dashboard: show refresh button in kiosk mode (#17032)
  Devenv: Updated gauge test dashboard
  Chore: reintroduce gosec (#17021)
  Gauge: tweaks to background color and height usage (#17019)
  Feature: provide multildap server configuration (#16914)
  (feat/explore): Support for new LogQL filtering syntax (#16674)
  fix(explore): Prevent double querying for Prometheus and Loki (#17004)
  Chore: No implict any fixes (#17020)
  move log package to /infra (#17023)
  Chore: Lowered implicit anys limit to 5386
  Chore: Updated snapshot
  Select: Fixed isOpen issue
  Chore: Typescript no-implicit any fixes progress (#17018)
  GraphPanel: show results for all SeriesData (#16966)
  Fix: Wrap value of multi variable in array when coming from URL (#16992)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master: (142 commits)
  Build: specify build flag for `docker-compose up` (#17411)
  Add a @grafana/runtime package with backendSrv interface (#16533)
  Database: Initialize xorm with an empty schema for postgres (#17357)
  docs: configuring custom headers in the dataproxy (#17367)
  Explore: Queries the datasource once per run query and uses DataStreamObserver (#17263)
  Feature: Adds redux action logging toggle from url params (#17368)
  Build: Adds e2e tests back to master workflow with better error messages and with artifacts (#17374)
  Explore: Handle datasources with long names better in ds picker (#17393)
  Annotations: Improve annotation option tooltips (#17384)
  InfluxDB: Fixes single quotes are not escaped (#17398)
  Chore: Bump axios to 0.19.0 (#17403)
  Alerting: golint fixes for alerting (#17246)
  Batch disable users (#17254)
  Chore: Remove unused properties in explore (#17359)
  MySQL/Postgres/MSSQL: Add parsing for day, weeks and year intervals in macros (#13086)
  Security: Prevent csv formula injection attack  (#17363)
  LDAP: remove unused function (#17351)
  Enterprise: remove gofakeit dep (#17344)
  Explore: Update time range before running queries (#17349)
  Build(package.json): improve npm commands (#17022)
  ...

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  Provisioning: Support folder that doesn't exist yet in dashboard provisioning (#17407)
  Codestyle: Fix govet issues (#17178)
  @grafana/runtime: expose location update (#17428)
  Fix: Adds context to list of keys that are not part of query (#17423)
  Prometheus: Correctly escape '|' literals in interpolated PromQL variables (#16932)
  Explore: Makes it possible to use a different query field per mode (#17395)
  DataSourceApi: remove ExploreDataSourceApi (#17424)
  Fix: change angular loader paths (#17421)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  Cloudwatch: Add AWS DocDB metrics (#17241)

ryantxu committed:
merge master

ryantxu committed:
move files to data package

ryantxu committed:
not used

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master: (156 commits)
  DataFrame: convert from row based to a columnar value format (#18391)
  Packages: Temporarily skip canary releases if packages build fail (#18577)
  Update latest.json to latest stable version (#18575)
  Docs: Update changelog for v6.3.3 (#18569)
  Graph: Fixed issue clicking on series line icon (#18563)
  grafana/toolkit: Unpublish previous "next" version when releasing a new one (#18552)
  Toolkit: write PR report to a folder with the circle build number (#18560)
  CI: Fail build if yarn.lock is not up to date (#18555)
  Chore: Updates react-dependant packages to address react warnings (#18549)
  Prometheus: Fix regression of rerunning query on legend/interval change (#18147)
  Explore/Prometheus: More consistently allows for multi-line queries (#18362)
  Login: Fixes undefined redirect (#18545)
  Plugins: expose react-redux, redux (#18501)
  TimeSeries: assume values are all numbers (#18540)
  Login: Angular to React (#18116)
  InfoTooltip: Info icon with tooltip (#18478)
  Annotations: Fix failing annotation query when time series query is cancelled (#18532)
  remotecache: support SSL with redis (#18511)
  QueryData: Handle that response data must be array (#18504)
  React: Rename deprecated UNSAFE_componentWillReceiveProps (#18526)
  ...

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  Docker: switch docker image to alpine base with phantomjs support (#18468)
  Backend: Adds support for HTTP/2 (#18358)
  Explore: Fixes error when switching from prometheus to loki data sources (#18599)
  TimePicker: Set time to to 23:59:59 when setting To time using calendar (#18595)
  Prometheus: Return labels in query results (#18535)
  Docs: Update changelog and docs for annotation region change (#18593)
  Refactor: move KeyValue and deprecation warning to @grafana/data (#18582)
  Annotations: use a single row to represent a region (#17673)
  Docs: Update upgrading guide (#18547)
  Docs: Adds tests requirement to bugs checklist (#18576)

ryantxu committed:
update to columnar

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  SingleStat: use DataFrame results rather than TimeSeries/TableData (#18580)
  TestData: attach labels to series results (#18653)
  Singlestat: Disable new singlestat gauge usage (#18610)
  Explore: Fixes query field layout in splitted view for Safari browsers (#18654)
  MSI: new long file names are causing error building MSI (#18646)
  Auth: change the error HTTP status codes (#18584)
  Refactor: EmptyListCTA (#18516)
  Build: Upgrade to go 1.12.9 (#18638)
  Chore: Revert React 16.9.0 bump (#18634)
  Azure Monitor and Log Analytics converted and separated into components (#18259)
  Rewrite user profile edit to react (#17917)
  Docs: remove codecov badge (#18623)
  Prometheus: Prevents panel editor crash when switching to Prometheus datasource (#18616)
  Chore: Rename Popper to Popover (#18543)
  SingleStat: add a gauge migration call to action button in the editor (#18604)
  Build: update revive dependency (#18585)
  LDAP: multildap + ldap integration (#18588)

dprokop committed:
Merge branch 'master' into transforms

dprokop committed:
Add more tests for nameMatcher

dprokop commented on packages/grafana-data/src/utils/matchers/matchers.ts:
```diff
@@ -0,0 +1,40 @@
+import { Field, DataFrame } from '../../types/dataFrame';
+import { Registry, RegistryItemWithOptions } from '../registry';
+
+export type DataMatcher = (series: DataFrame, field?: Field) => boolean;
```
@ryan going through the code. Seems like I get the feeling what what we are about to achieve with this concept.

One thing thad bugs me here is what is your thinking behind pushing series as an argument of a matcher?

ryantxu replied:
Perhaps we should have a FieldMatcher and a FrameMatchet?

dprokop committed:
Fix invert predicate

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  Streaming: improve JSDocs for DataSourceAPI streaming support (#18672)

ryantxu committed:
add fluent API

ryantxu committed:
remove calc snapshot

ryantxu committed:
split Field matchers and Frame matchers

ryantxu committed:
split filter transformers

ryantxu commented:
@dprokop -- since you have looked at this I have:
1. renamed `calc` to `reduce`
2. split `matchers` into `FieldMatcher` and `FrameMatcher` -- this causes a bunch of duplication, but I think the distinction is better than doing both in one pass
3. removed the super early UI component to pick a matcher -- lets do that in a different issue
4. added a super basic api to `dataTransformers` to give an idea on how we could more easily call this from javascript

I think this is almost ready to commit and we can work on the UI and maybe api in follow on issues

dprokop commented:
@ryantxu thanks, will have a look soon!

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  QueryEditor: check if optional func toggleEditorMode is provided (#18705)
  Emails: remove the yarn.lock (#18724)
  OAuth: Support JMES path lookup when retrieving user email (#14683)
  Emails: resurrect template notification (#18686)
  Email: add reply-to and direct attachment (#18715)
  Dashboard: Adds Logs Panel (alpha) as visualization option for Dashboards (#18641)
  Heatmap: Add Cividis and Turbo color schemes (#18710)
  Units: add counts/sec (cps) and counts/min (cpm) in Throughput (#18702)
  Dev Docker: Use golang:1.12.9-alpine to prevent glibc mismatch. (#18701)
  Docs: Fix broken link for the Grafana on RHEL or Ubuntu tutorial (#18697)
  Fixes several usability issues with QueryField component  (#18681)
  convert teams section of user profile to react (#18633)
  Singlestat/Gauge/BarGauge: Improvements to decimals logic and added test dashboard (#18676)
  Emails: Change text (#18683)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into transforms

* grafana/master:
  @grafana/data: improve the CircularVector api (#18716)

dprokop committed:
Merge branch 'master' into transforms

dprokop committed:
Fix typo

dprokop commented:
@ryantxu went through the entire thing and it feels like it's ok to be merged. Not sure I understand the need for :

> added a super basic api to dataTransformers to give an idea on how we could more easily call this from javascript

correctly. You want the transformers to be available for devs directly as well? I was thinking more of a reusable UI for that. But after second though, these are/going to be some basic operations on data, so maybe having access to those (i.e. for plugins) makes sense.

I think we gonna need sth similar for matchers registry as well, as otherwise we there will be need fo using matcher's ID when passing this to i.e. reduce transformer. But we can work on that when working on #18740

torkelo reviewed (approved):
Looks good!
